tools/wptrunner/wptrunner/browsers/wktr.py

```diff
@@ -0,0 +1,239 @@
+
+        self._stdout_queue = Queue()
+        self._stderr_queue = Queue()
+        self._stdin_queue = Queue()
+
+        self._stdout_reader = self._create_reader_thread(self._proc.stdout, self._stdout_queue)
+        self._stderr_reader = self._create_reader_thread(self._proc.stderr, self._stderr_queue)
+        self._stdin_writer = self._create_writer_thread(self._proc.stdin, self._stdin_queue)
+
+        # WebKitTestRunner is likely still in the process of initializing. The actual waiting
+        # for the startup to finish is done in the WKTRProtocol.
+        self.logger.debug("WebKitTestRunner has been started.")
+
+    def stop(self, force=False):
+        self.logger.debug("Stopping WebKitTestRunner...")
+
+        if self.is_alive():
+            kill_result = self._proc.kill(timeout=5)
+            # This makes sure any left-over child processes get killed.
+            # See http://bugzilla.mozilla.org/show_bug.cgi?id=1760080
+            if force and kill_result != 0:
+                self._proc.kill(9, timeout=5)
+
+        # We need to shut down these queues cleanly to avoid broken pipe error spam in the logs.
+        self._stdout_reader.join(2)
+        self._stderr_reader.join(2)
+
+        self._stdin_queue.put(None)
+        self._stdin_writer.join(2)
+
+        for thread in [self._stdout_reader, self._stderr_reader, self._stdin_writer]:
+            if thread.is_alive():
+                self.logger.warning("WebKitTestRunner IO threads did not shut down gracefully.")
+                return False
+
+        stopped = not self.is_alive()
+        if stopped:
+            self._proc = None
+            self._stdout_queue.close()
+            self.logger.debug("WebKitTestRunner has been stopped.")
+
+            # We sometimes accumulate too many process-related objects,
+            # ultimately running out of OS file handles, via circular
+            # references to them, thus manually trigger a GC while stopping.
+            gc.collect()
+        else:
+            self.logger.warning("WebKitTestRunner failed to stop.")
+
+        return stopped
+
+    def is_alive(self):
+        return self._proc is not None and self._proc.poll() is None
+
+    def pid(self):
+        return self._proc.pid if self._proc else None
+
+    def executor_browser(self):
+        """This function returns the `ExecutorBrowser` object that is used by other
+        processes to interact with WebKitTestRunner. In our case, this consists of the three
+        multiprocessing Queues as well as an `io_stopped` event to signal when the
+        underlying pipes have reached EOF.
+        """
+        return ExecutorBrowser, {"stdout_queue": self._stdout_queue,
+                                 "stderr_queue": self._stderr_queue,
+                                 "stdin_queue": self._stdin_queue}
+
+    def check_crash(self, process, test):
+        return not self.is_alive()
+
+    def _create_reader_thread(self, stream, queue):
```

Comment:
Although I am mildly surprised that you can't use mozprocess's output handling, which iirc is pretty similar to what you have here.